domain/src/test/java/org/oppia/domain/audio/AudioPlayerControllerTest.kt

```
@@ -0,0 +1,465 @@
+    assertThat(audioPlayerResultCaptor.value.isSuccess()).isTrue()
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().type).isEqualTo(PlayStatus.PAUSED)
+  }
+
+  @Test
+  fun testAudioObserver_invokePrepared_capturesCorrectProgress() {
+    arrangeMediaPlayer()
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().position).isEqualTo(0)
+  }
+
+  @Test
+  fun testAudioObserver_invokeSeekTo_capturesCorrectProgress() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.seekTo(500)
+    audioPlayerController.play()
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+  }
+
+  @Test
+  fun testAudioObserver_invokePlay_capturesCorrectDuration() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.play()
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().duration).isEqualTo(1000)
+  }
+
+  @Test
+  fun testAudioObserver_invokeChangeDataSource_capturesCorrectProgress() {
+    arrangeMediaPlayer()
+
+    audioPlayerController.seekTo(500)
+    audioPlayerController.changeDataSource(TEST_URL2)
+    shadowMediaPlayer.invokePreparedListener()
+    audioPlayerController.play()
+
+    verify(mockAudioPlayerObserver, atLeastOnce()).onChanged(audioPlayerResultCaptor.capture())
+    assertThat(audioPlayerResultCaptor.value.getOrThrow().position).isEqualTo(0)
+  }
+
+  @Test
+  fun testAudioObserver_invokeRelease_checkRemoveObservers() {
+    val playProgress = audioPlayerController.initializeMediaPlayer(TEST_URL)
+    playProgress.observe(fragment, mockAudioPlayerObserver)
+
+    audioPlayerController.play()
+    audioPlayerController.releaseMediaPlayer()
+
+    assertThat(shadowMediaPlayer.state).isEquivalentAccordingToCompareTo(ShadowMediaPlayer.State.END)
```

Prefer using isEqualTo() when comparing enum values. The autocomplete in Android Studio inadvertently defaults to the wrong method for this case.

The 'compareTo' is more useful when verifying Comparable logic, but for enums this automatically and correctly matches equals().